File: backendEcomm/src/models/user_model.js
const { Schema, model } = require('mongoose');
const uuid = require('uuid');
const bcrypt = require('bcrypt');

// Define the user schema
const userSchema = new Schema({
    id: { type: String, unique: true },
    fullName: { type: String, default: "" },
    email: { type: String, unique: true, required: true },
    password: { type: String, required: true },
    phoneNumber: { type: String, default: "" },
    address: { type: String, default: "" },
    city: { type: String, default: "" },
    state: { type: String, default: "" },
    profileProgress: { type: Number, default: 0 },
    profileImage: { type: String, default: "" },
    updatedOn: { type: Date },
    createdOn: { type: Date }
});

// Pre-save hook to generate UUID and hash the password
userSchema.pre('save', function(next) {
    this.id = uuid.v1();
    this.updatedOn = new Date();
    this.createdOn = new Date();

    const salt = bcrypt.genSaltSync(10);
    this.password = bcrypt.hashSync(this.password, salt);

    next();
});

// Pre-update hook to update the `updatedOn` field
userSchema.pre(['update', 'findOneAndUpdate', 'updateOne'], function(next) {
    const update = this.getUpdate();
    delete update._id;
    delete update.id;

    this.updatedOn = new Date();

    next();
});

// Create the model
const UserModel = model('User', userSchema);

module.exports = UserModel;

File: app/lib/data/models/user/user_model.dart
class UserModel {
  String? sId;
  String? fullName;
  String? email;
  String? password;
  String? phoneNumber;
  String? address;
  String? city;
  String? state;
  int? profileProgress;
  String? id;
  String? updatedOn;
  String? createdOn;
  String? profileImage; // New field for profile image URL or path

  UserModel({
    this.sId,
    this.fullName,
    this.email,
    this.password,
    this.phoneNumber,
    this.address,
    this.city,
    this.state,
    this.profileProgress,
    this.id,
    this.updatedOn,
    this.createdOn,
    this.profileImage,
  });

  UserModel.fromJson(Map<String, dynamic> json) {
    sId = json['_id'] ?? ''; // Default empty if null
    fullName = json['fullName'] ?? '';
    email = json['email'] ?? '';
    password = json['password']; // Allow null
    phoneNumber = json['phoneNumber'] ?? '';
    address = json['address'] ?? '';
    city = json['city'] ?? '';
    state = json['state'] ?? '';
    profileProgress = json['profileProgress'] ?? 0; // Default to 0
    id = json['id'] ?? '';
    updatedOn = json['updatedOn'] ?? '';
    createdOn = json['createdOn'] ?? '';
    profileImage = json['profileImage'] ?? '';
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': sId,
      'fullName': fullName,
      'email': email,
      'password': password,
      'phoneNumber': phoneNumber,
      'address': address,
      'city': city,
      'state': state,
      'profileProgress': profileProgress,
      'id': id,
      'updatedOn': updatedOn,
      'createdOn': createdOn,
      'profileImage': profileImage,
    };
  }
}
